### DISCUSSION TITLE
Development focus

### DISCUSSION THREAD

**Discussion by HenrikJannsen** [bisq-network/bisq#6747] [General]
There are a few new devs (@AllyKaz, @napoly) around (which is great!) and some work is done in the testing and API area. I would like to discuss what Bisq is intended to focus on dev-wise. 

Bisq has suffered in the past from ill-invested money in work which never brought improvements to the project (in some cases even the opposite). That is mainly because the dev area is largely un-managed and new devs are often left alone to figure out what to work on. Communication in the dev area is also very poor unfortunately (everyone too buys with coding and no one wants to play the managing role).

So here some suggestions how we can avoid that from happening again:
- Make clear where we want to invest money and focus development in Bisq 1
- Make clear that Bisq 2 is in development which aims to take over Bisq 1 once the Multisig protocol is implemented there. This might take another 6 months or so, but investment in Bisq 1 should be seen in that context.
- Try to improve communications in the dev area. 

Here are my personal priorities and view on what Bisq should focus on and what not:

The Bisq 1 API has been a failure from my point of view. We still do not know if and how many people are using it but from feedback from several devs who wanted to use it (and most gave up after a while) it seems its too cumbersome to use. I have never used it myself, so that is all second hand information. Bisq invested a lot of money in it with more then 1 year dev work. We have to ensure to not throw more money into that grave anymore. 

Investment in the test coverage has a similar track record. Bisq invested lot of money in the past in devs who wanted to improve that and in some parts they got for sure improvements but overall the invested money was not justified by the results. In some cases it made the code base worse by adding more boilerplate code without any real improvements (like passing the Clock object everywhere). 

Bisq is very different to traditional IT industry projects and need to be treated with that in mind. Also our resources are limited and to get a valuable test environment is a huge effort Bisq simply cannot afford. So what often makes perfect sense in a traditional IT industry project often does not work for Bisq.

I am not talking about pure unit tests testing some limited behavior but about the complex integration tests. Simple unit tests are of course welcome but that is usually not where the problems happens.
The bugs are mostly caused in the highly asynchronous and often multiple domains spanning areas often involving UI. To test that requires very complex mock structures and integration tests. In the P2P network one dev implemented that to some extent, but even the quality of that was very high it turned out that it sometimes did not test the real behavior and the highly complex mock structures caused a huge burden for future changes. A small change code-wise often triggered then a day work to get the tests working again. Not because the change broke something but because the mock structure did not represent the real behavior and thus the change could not be easily integrated. 

I think dev on Bisq 1 should be focused on areas where the user gets a direct benefit not in dev infrastructure (also with the limited life-span of Bisq 1 in mind). 
Those are features requested by users or by the support team, bug fixes, UI improvements and performance optimizations.

The work done by @jmacxx and @stejbac (he did a lot of amazing performance optimizations) are good examples for that.

Those are my personal views and I invite any reader to post their view. I hope we can come to some consensus on what we want to focus on to avoid disappointments for new developers and ill-invested expenses for the Bisq DAO.

**Comment by pazza83** [comment]
Thanks for creating the post.

When looking to decide where to focus development I think it is important to consider them in the context of the current opportunities and threats.

**Opportunities** 

- Increased interest of people to consider P2P / No KYC / decentralization etc, alongside a general interest, awareness and confidence of people in bitcoin overtime.
- Reduced number of P2P platforms means there has been consolidation of the market (Local Bitcoins, Paxful and Local Cryptos have all gone). 

**Threats**

- Potential regulation of P2P platforms. 
- Spending increasing longer durations in high fee environments where the Bisq 1 (multisig protocol) is not economically viable.
- Competition from other p2p and No KYC platforms, and potential atomic swap protocols, that take potential users away from Bisq. 57% of Bisq's revenue comes from XMR trades (last year 2022 and year to date). 
- As bitcoin increases in price less total bitcoin volume is traded on the platform reducing the revenue in bitcoin denominated terms.
- CBDCs, I have concerns that when these are implemented it will be easier for centralized entities to target P2P traders. You only have to look at how services like Wise or Revolut prevent users that trade any significant volume on Bisq to see how easy it would be for a centralized entity to target (block, reverse, de-bank etc) Bisq traders reliant on using CBDCs for trading. For myself, I see Bisq predominantly as a platform for buying and selling bitcoin for fiat, but maybe with the coming of CBDCs Bisq should lean in more to trading of alts and who knows in the future even goods / services for bitcoin.

For the short to medium term I think Bisq should focus it's development on things that strengthen the DAO. Those being continue to decentralize as much as possible and look to increase revenue by increasing the trade volume in bitcoin terms. 

I see having a strong DAO as; increased contributor and user participation, increased BSQ price, and increased BSQ market cap. I think these things would be by and large the result that come from an increase in trading volume and an improved user experience. Achieving a strong DAO should also help with one of Bisq's limitations, a lack of dev resources.

I think one of the DAO's major strengths is the level of decentralization. So I have nothing to suggest to improve this other than I see it as something that should continue to take priority over platform development (a good example of this would be the decentralization of the burning man role).

My thoughts regarding dev priorities that would increasing revenue the trade volume are:

**Address the issue that will be caused with spending longer periods in high fee environments**

- Dedicated protocol in Bisq 2 specifically for Lightning trades (does not have to start with a [escrow method](https://github.com/bisq-network/proposals/issues/416) - can be as simple as users choosing to send receive using Lightning rather than on-chain, same way as Bisq easy will work).
- Reduce multisig trade protocol to [one single transaction](https://github.com/bisq-network/proposals/issues/279). There will always be a user base for on-chain trades. If we can reduce the number of transactions needed then this should be a priority. 
- Lighting escrow trade protocol once feasible.

**Address the issue that will be caused with treats to Bisq's XMR trade volume (other p2p platforms and coming increase in atomic swap usage)**
 
XMR trades generate 57% of Bisq's revenue but I do not think much if any percentage of dev resources go into developing specific improvements for XMR traders. I think Bisq should use more of it's resources to make improvements for XMR traders. This might bring with it new contributors.  

- Dedicated protocol in Bisq 2 specifically for XMR to fiat trades (Again it does not have to start with a multisig. Just replicate the bitcoin <-> fiat Bisq Easy protocol for XMR traders. 
- Overtime integrate a multisig protocol for XMR users in Bisq 2 that can allow trades using XMR as the security deposit. 
- Progress to an Atomic Swap market place when feasible
- Potentially doing something similar with Eth (eg Eth to fiat on Bisq Easy) might bring more users and should be a quick win for increasing revenue and growing user base.

**Prioritize what protocols should be developed for addition to Bisq 2 based on the likely revenue they will generate**

There are lots of great protocols that could be added to Bisq 2 and I think it is important to evaluate them in terms of what will increase the revenue the most. Certain protocols might be technically great but will not have much adoption. Other protocols might not offer much in the way of innovation but will be used a lot. IMHO these should take priority. 

Long term goals would be to increase accessibility by having Bisq accessed via a mobile, or web app in a standalone decentralized manor. I think this is a long way off though and will be achieved with improvements overtime with technology rather than anything Bisq needs to focus on currently.

**Comment by salokiam** [comment]
Onramping new devs is very difficult at Bisq. We have seen many devs coming by with a lot of enthusiasm, but dropping out shortly after. The codebase is not so easy to understand, that you give it to someone and expect that he will be able to implement a feature. Giving them some guidance and some areas they can work on would be essential because when you come to a new project, you probably dont see what's needed.

And then there is the problem of misguided dev-effort as you pointed out. That hurts the project double, not only the money spend for little or nothing, but also the opportunity missed to have something important to be implemented.
Having experience with internal teams as well as distributed teams, I would be interested to take on such a role. I will DM you.
Regarding the aspect of where we should focus on, this (traditionally) is a complex discussion with quite a number of people. Once there is a decision on what we want to spend the money on, that role could take care that we get what we needed.

But to summarize it, I think you have hit the nail on the head concerning the problems why we don't have more devs and effectiveness in development.

**Comment by HenrikJannsen** [comment]
Regaring protocols for Bisq 2. There have been discussions about a LN->Liquid->LN protocol based on a suggestion from Adam Back. I think that is the most feasible solution and the concept could be used  also for XMR. Let me explain...

2 traders want to trade LN-BTC to fiat and use the Bisq Multisig protocol in the background.
1. They first move their LN-BTC via reverse submarine swap to Liquid-BTC (L-BTC). 
2. Then they do the Multisig protocol on Liquid (1 min confirmation time, super low fees, better privacy due confidential transactions)
3. After that is completed the L-BTC gets back to LN-BTC via a submarine swap

So it is a chain of 3 trades/protocols. 
1. Reverse submarine swap LN-BTC -> L-BTC
2. Multisig protocol on Liquid
3. Submarine swap L-BTC -> LN-BTC

The swaps are conceptually similar as atomic cross chain swaps, so they are to a very high degree secure and can be automated if there is enough liquidity from market makers. If liquidity would be a problem we could use a centralized provider like Bolz.exchange (but I think that is not needed).
Once we have XMR atomic cross chain swaps we could do the same for XMR:
XMR -> L-BTC, Multisig on Liquid, L-BTC -> XMR.

One can think of it like entering a casino and converting fiat to casino coins. Then one can use those coins for the "services" of the casino. Once done, they convert it back to fiat. 

Doing multisig in Monero (like Haveno is doing it) is very problematic as multisig in Monero is much more complicate as in Bitcoin and every tx needs quite some time before it can be used in the next transactions, so you cannot create a protocol based on chained transactions without long waiting time in between (I guess its 20-30 min as far I remember). So UX will suck and complexity is very high.

My personal roadmap for protocols on Bisq 2 is: 
1. Bisq Easy 
2. Bisq Multisig with improvements (staged payout txs so that non responding trader cases don't need to go to arbitration, remove fee txs, maybe 1 tx protocol (but I doubt that is really feasible - so I would count with 2 txs)
3. Submarine swaps and reverse Submarine swaps (I guess they are technically very similar)
4. Port Bisq Multisig to Liquid
5. Chained LN -> L-BTC -> LN protocol as described above
6. Maybe: Liquid atomic swaps (Tether-BTC)
7. XMR atomic swaps
8. XMR -> L-BTC -> XMR chained protocol as described above

BSQ swaps could be added at any moment, but I guess that's not the most pressing.
The escrow based LN trade protocol from @stejbac is another candidate but I am not sure if we should work on both LN protocol versions as it would be kind of waste of resources without a big benefit for the user as both are a bit similar from security point of view.

If Bisq Easy turns out to work well the reputation based approach could be developed further for other currencies and more reputation sources. It could turn out to be a very flexible solution. But lets see first how it gets adopted by users.

Beside the protocol development I see those dev areas as high prio:

1. Web interface (@salokiam is working on that) to enable to use Bisq 2 on headless nodes (Umbrel,...)
2. Rest API for trading bots
3. Android mobile app for Bisq Easy (See https://github.com/bisq-network/bisq2/discussions/298)
4. Maybe iOS app, but that is more limited due Apples restrictions. Cannot be done as a full node but will require some relay node model. Not sure if its worth the effort and risk that Apple kills it at any moment.

On Bisq 2 there are a few open areas where we would need highly skilled and experienced devs:
1. Make I2P integration production ready
2. Launcher app model (See https://github.com/bisq-network/bisq2/issues/200)
3. Rest API
4. Android app

There are a long list of open issues here: https://github.com/bisq-network/bisq2/issues but most will be covered by myself or @alvasw but we will not have the bandwidth for the larger tasks mentioned above. So if any new dev is interested in one of those they are welcome. 

I am aware that those are not trivial tasks and require a high level of experience and skill set.
I fear working on Bisq comes with a rather high bar.

Here would be my summary a dev should bring when working at Bisq:
- Multiple years of Java experience as senior developer
- At least 20 hours per week availability, better 40 hours
- Experience in at least one of those fields: Network layer, Bitcoin dev, LN dev, cryptography, software architecture, protocol development, mobile development, UI development.
- Ability to work autonomous to a high degree. 
- Ability to figure out problems and develop solutions. 
- Good communication skills (ok, that is not the biggest strength of the dev team I have to admit ;-))  

On Bisq 1 I do not see a large space for new devs. Work on Bisq 1 is difficult as the code base is highly complex and has quite some "wear" and it takes considerable time to get familiar with it.
The larger improvements like a protocol change with a staged payout transaction are IMO better invested in Bisq 2. Mobile app is unfeasible for Bisq IMO. LN would require a different trade protocol which again is better invested in Bisq 2.

Areas where it might make sense are:
- UI/UX improvments (dev need to have JavaFX skills)
- Bug fixes
- New payment methods 


My 5 cents....

**Reply by MwithM** [reply]
> If Bisq Easy turns out to work well the reputation based approach could be developed further for other currencies and more reputation sources. It could turn out to be a very flexible solution. But lets see first how it gets adopted by users.

Flexibility should be the key for Bisq2. 
The way I understand it, Bisq Easy already could allow to trade any pair (like XMR/liquid) with a few UX changes. Letting users to trade any asset at would bring many users, and it has almost no extra cost. In my opinion, the sooner Bisq Easy it's open to a wider range of assets other than fiat/btc, the better.

Let's see how the cloning an offer feature works out, I'm very optimistic with it and I think it will generate higher volume, specially for payment methods with lower volume. These are the kind of features that make the difference, but I also understand it's been thought for a long time until you found a viable way to implement it that only looks easy in retrospective.

**Reply by HenrikJannsen** [reply]
Yes in Bisq Easy one could trade anything to anything. But to stay focussed on the target user group of newbies who want to get into Bitcoin we decided to stick with Fiat/BTC in the UI. Adding more options (altcoin-fiat) would add friction and confusion to the user group we target. But sure later once we got some experience how its accepted by users we can consider to add for instance XMR-Fiat or BTC-altcoin....

**Reply by counterweightoperator** [reply]
I agree with the idea that, even if extending the Bisq Easy protocol for other asset pairs is simple code wise, it comes with a cost in terms of UX. Perhaps this cost can be reduced by keeping a UX path focused on Bitcoin that funnels users with little experience in the right way, but allows advanced users to step out of the "usual" path.

**Comment by alvasw** [comment]
I suggest creating a small document with a list of tasks and link to it from the README file. We already have several good tasks for new contributors.

#### Bisq 1 API
IMO we should focus too much on the Bisq 1 API anymore. I used to it to build few trading bots, and it's not well-designed. I have to admit that I ignored the API module when I fixed threading bugs in the desktop app. It was too cumbersome. The only reason to justify to invest more into the API is if someone is using it production.

#### Tests
I usually write the tests before implementing the code, and it's not a secret that I like tests. However, when @napoly upgraded JUnit, I had to look through most tests in Bisq 1. It's a mess and the existing tests are too concrete and make the implementation code worse. Little changes break the tests and make them useless.

#### Focus for Bisq 1
We should only work on things that benefit our users directly and fix bugs.

#### Focus for Bisq 2
Henrik's roadmap for protocols on Bisq 2:
1. Bisq Easy
2. Bisq Multisig with improvements (staged payout txs so that non responding trader cases don't need to go to arbitration, remove fee txs, maybe 1 tx protocol (but I doubt that is really feasible - so I would count with 2 txs)
3. Submarine swaps and reverse Submarine swaps (I guess they are technically very similar)
4. Port Bisq Multisig to Liquid
5. Chained LN -> L-BTC -> LN protocol as described above
6. Maybe: Liquid atomic swaps (Tether-BTC)
7. XMR atomic swaps
8. XMR -> L-BTC -> XMR chained protocol as described above

**Comment by HenrikJannsen** [comment]
> I usually write the tests before implementing the code, and it's not a secret that I like tests. However, when @napoly upgraded JUnit, I had to look through most tests in Bisq 1. It's a mess and the existing tests are too concrete and make the implementation code worse. Little changes break the tests and make them useless.

Yes you are one of the very few example who write test which totally make sense and even highly complex integration tests. But that is an art not many developers are mastering. I agree totally to your statements!


I just had a conversation with @salokiam and he will start to manage those issues (improve/update onboarding documentation and be a point for contact for new devs).

Here are a few more open tasks for Bisq 2 I forgot in the above list:
- Research for the best LN library to be used for LN trade protocols. I am sure @stejbac has gained a lot of experience already in that regards.
- Wallet integration in Bisq 2. @alvasw implemented already Bitcoin Core, Electrum and Elements. But Electum as light-mode wallet will be too restricted for more complex trade protocols, so we consider to replace that by some other solution. BDK or Sparrow wallet might be candidates. As @alvasw is very busy with many other things that might be an area where support from someone who has experience with Bitcoin libraries would be welcome as well.

**Comment by salokiam** [comment]
I am looking at your [roadmap for protocols](https://github.com/bisq-network/bisq/discussions/6747#discussioncomment-6355645) for Bisq2. Given the current development speed (Bisq2 is in development for 2 years, Bisq Easy has probably used 1 year of 1 dev), I estimate it will take several years until we have implemented Lightning.
Point 1 (Bisq Easy), could probably finish soon.
But Point 2, Bisq Multisig with improvements. Since we havent even figured out the concept (and current dev speed) this might easily take 2 years. Then we are ready to tackle lightning. 

Until lightning is done, we havent even created anything where we can improve the revenue situation. As Bisq Easy wont generate anything and MS is only a replacement for existing, but not generating substantially more users or revenue.

Lightning being a very complex, i would think (at current development speed) this will take another 2 years minimum. That makes 4 years until we have a protocol with can drive more adoption of Bisq.

From talking to potential users, I regularly hear the following requests:
1. Mobile
2. Lightning
3. Improved UX

Since my comment here is only about developing protocols, Lightning is more important than an improved MultiSig. If I find time, I will address mobile in a different post.

So upshot of this little article is to change the ordering of your prios for protocol development. After Bisq Easy, go for Lightning and then for improved MultiSig.

**Comment by HenrikJannsen** [comment]
The Lightning to Fiat protocol is using the ported Multisig protocol to Liquid. The other option would be the Lightning protocol concept from @stejbac .

To the dev time estimations: Yes if we stick with current limited dev resources it will take some time. But hey there are many devs out there, so lets get some work done. If we would have 2-4 more fulltime devs things would look very different.
